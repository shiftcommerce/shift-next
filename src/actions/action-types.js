// Action types
export const AUTOFILL_ADDRESS = 'AUTOFILL_ADDRESS'
export const AUTOFILL_BILLING_ADDRESS = 'AUTOFILL_BILLING_ADDRESS'
export const BILLING_ADDRESS_CREATED = 'BILLING_ADDRESS_CREATED'
export const CARD_ERRORS = 'CARD_ERRORS'
export const CART_UPDATED = 'CART_UPDATED'
export const CREATE_ACCOUNT = 'CREATE_ACCOUNT'
export const CREATE_ORDER = 'CREATE_ORDER'
export const ERROR_ACCOUNT = 'ERROR_ACCOUNT'
export const ERROR_LOGIN = 'ERROR_LOGIN'
export const ERROR_MENU = 'ERROR_MENU'
export const ERROR_ORDER = 'ERROR_ORDER'
export const ERROR_REGISTRATION = 'ERROR_REGISTRATION'
export const CLEAR_ACCOUNT_ERRORS = 'CLEAR_ACCOUNT_ERRORS'
export const GET_ACCOUNT = 'GET_ACCOUNT'
export const GET_LOGIN = 'GET_LOGIN'
export const GET_MENU = 'GET_MENU'
export const GET_PRODUCT = 'GET_PRODUCT'
export const GET_CUSTOMER_ORDERS = 'GET_CUSTOMER_ORDERS'
export const PASSWORD_RESET = 'PASSWORD_RESET'
export const REQUEST_CARD_TOKEN = 'REQUEST_CARD_TOKEN'
export const SET_PRODUCT = 'SET_PRODUCT'
export const SET_ACCOUNT = 'SET_ACCOUNT'
export const SET_LOGGED_IN = 'SET_LOGGED_IN'
export const SET_LOGIN = 'SET_LOGIN'
export const SET_CUSTOMER_ORDERS = 'SET_CUSTOMER_ORDERS'
export const SET_ADDRESS = 'SET_ADDRESS'
export const SET_ADDRESS_BOOK_ENTRY_BILLING = 'SET_ADDRESS_BOOK_ENTRY_BILLING'
export const SET_ADDRESS_BOOK_ENTRY_SHIPPING = 'SET_ADDRESS_BOOK_ENTRY_SHIPPING'
export const SET_ADDRESS_BOOK = 'SET_ADDRESS_BOOK'
export const SET_CARD_TOKEN = 'SET_CARD_TOKEN'
export const SET_ORDER = 'SET_ORDER'
export const DELETE_ADDRESS = 'DELETE_ADDRESS'
export const SHIPPING_ADDRESS_CREATED = 'SHIPPING_ADDRESS_CREATED'
export const SET_CHECKOUT_INPUT_VALUE = 'SET_CHECKOUT_INPUT_VALUE'
export const SET_CHECKOUT_INPUT_VALIDATION_MESSAGE = 'SET_CHECKOUT_INPUT_VALIDATION_MESSAGE'
export const SET_CHECKOUT_INPUT_SHOWN = 'SET_CHECKOUT_INPUT_SHOWN'
export const SET_MENU = 'SET_MENU'
export const SET_PAYMENT_ERROR = 'SET_PAYMENT_ERROR'
