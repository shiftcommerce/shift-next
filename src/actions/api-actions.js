// Libs
import ApiClient from '../lib/api-client'
import JsonApiParser from '../lib/json-api-parser'

export const readEndpoint = (request) => {
  return (dispatch) => {
    dispatchInitialAction(dispatch, request)
    return new ApiClient().read(request.endpoint, request.query)
      .then(processResponse(dispatch, request, [200, 304]))
      .catch((error) => {
        if (request.errorActionType) dispatch(setErroredTo(request.errorActionType, error, request))
      })
  }
}

export const postEndpoint = (request) => {
  return (dispatch) => {
    dispatchInitialAction(dispatch, request)
    return new ApiClient().post(request.endpoint, request.body)
      .then(response => {
        _determinePostDispatch(dispatch, request, response)
      })
      .catch(error => {
        if (request.errorActionType) dispatch(setErroredTo(request.errorActionType, error))
      })
  }
}

export const deleteEndpoint = (request) => {
  return (dispatch) => {
    dispatchInitialAction(dispatch, request)
    return new ApiClient().delete(request.endpoint, request.body)
      .then(processResponse(dispatch, request, [204]))
      .catch(error => {
        if (request.errorActionType) dispatch(setErroredTo(request.errorActionType, error))
      })
  }
}

function dispatchInitialAction (dispatch, request) {
  if (request.requestActionType) {
    dispatch(fetching(request.requestActionType, request.requestActionData))
  }
}

function sendResponse (actionType, parsedPayload) {
  return {
    type: actionType,
    payload: parsedPayload
  }
}

function fetching (actionType, actionData) {
  return {
    type: actionType,
    data: actionData
  }
}

function setErroredTo (actionType, data, request) {
  return {
    type: actionType,
    payload: {
      error: {
        data: data,
        request: request
      }
    }
  }
}

function _checkForErrors (response) {
  let validationPassed = true
  // Validation for customer account form
  if (response.data.length > 0) {
    response.data.forEach(item => {
      if (parseInt(item.status) === 422 || parseInt(item.status) === 404) {
        validationPassed = false
      }
    })
  }

  return validationPassed
}

function processResponse (dispatch, request, expectedStatusCodes) {
  return (response) => {
    if (expectedStatusCodes.includes(response.status)) {
      if (request.successActionType) {
        const parsedPayload = new JsonApiParser().parse(response.data)
        // TODO: remove this when all endpoints have been extracted into shift-api

        // If parsedPayload is empty or is an array of objects with only one key (id) it means it had already been parsed.
        let payload
        if (!parsedPayload || (Array.isArray(parsedPayload.data) && parsedPayload.data.every(e => Object.keys(e).length === 1))) {
          payload = response.data
        } else {
          payload = parsedPayload
        }
        dispatch(sendResponse(request.successActionType, payload))
      }
    } else {
      if (request.errorActionType) dispatch(setErroredTo(request.errorActionType, response.data, request))
    }
  }
}

function _determinePostDispatch (dispatch, request, response) {
  const validationPassed = _checkForErrors(response)

  if ((response.status === 202 || response.status === 201 || response.status === 200) && validationPassed) {
    if (request.successActionType) {
      const parsedPayload = new JsonApiParser().parse(response.data)
      // TODO: remove when all enpoints have been moved to shift-next
      const payload = parsedPayload || response.data
      dispatch(sendResponse(request.successActionType, payload))
    }
  } else {
    dispatch(setErroredTo(request.errorActionType, response.data))
  }
}
