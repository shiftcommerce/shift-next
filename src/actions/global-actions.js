import * as types from './action-types'

export function toggleLoading(state) {
  return {
    type: types.TOGGLE_LOADING,
    loading: state
  }
}
