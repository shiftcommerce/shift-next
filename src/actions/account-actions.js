// actionTypes
import * as types from '../actions/action-types'

// Actions
import { readEndpoint, postEndpoint } from './api-actions'

const accountRequest = {
  endpoint: '/getAccount',
  query: {},
  errorActionType: types.ERROR_ACCOUNT,
  requestActionType: types.GET_ACCOUNT,
  successActionType: types.SET_ACCOUNT
}

export function fetchAccountDetails (store) {
  return readEndpoint(accountRequest)
}

export function updateCustomerAccount (details) {
  const request = {
    endpoint: '/updateCustomerAccount',
    body: {
      ...details
    },
    successActionType: types.SET_ACCOUNT
  }

  return postEndpoint(request)
}

export function clearErrors () {
  return {
    type: types.CLEAR_ACCOUNT_ERRORS
  }
}

export function setLoggedInFromCookies () {
  return {
    type: types.SET_LOGGED_IN,
    payload: {}
  }
}

const customerOrdersRequest = {
  endpoint: `/customerOrders`,
  requestActionType: types.GET_CUSTOMER_ORDERS,
  successActionType: types.SET_CUSTOMER_ORDERS
}

export function getCustomerOrders () {
  return readEndpoint(customerOrdersRequest)
}

const forgottenPasswordRequest = (email) => {
  return {
    endpoint: '/forgotPassword',
    query: { email }
  }
}

export function requestPasswordResetEmail (email) {
  return readEndpoint(forgottenPasswordRequest(email))
}

export function passwordReset (token, password) {
  const request = {
    endpoint: '/passwordReset',
    body: {
      data: {
        type: 'password_recoveries',
        attributes: {
          token: token,
          password: password
        }
      }
    },
    successActionType: types.PASSWORD_RESET,
    errorActionType: types.ERROR_ACCOUNT
  }

  return postEndpoint(request)
}
