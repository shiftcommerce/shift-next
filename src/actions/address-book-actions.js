// actionTypes
import * as actionTypes from './action-types'

// Actions
import { readEndpoint, postEndpoint, deleteEndpoint } from './api-actions'

export function saveToAddressBook (address, options = {}) {
  const request = {
    endpoint: '/createAddressBookEntry',
    body: newAddressBookEntryPayload(address),
    successActionType: options.billing ? actionTypes.SET_ADDRESS_BOOK_ENTRY_BILLING : actionTypes.SET_ADDRESS_BOOK_ENTRY_SHIPPING
  }
  return postEndpoint(request)
}

export function fetchAddressBook () {
  const request = {
    endpoint: '/getAddressBook',
    successActionType: actionTypes.SET_ADDRESS_BOOK
  }
  return readEndpoint(request)
}

export function deleteAddressBookEntry (address) {
  const request = {
    endpoint: `/deleteAddress/${address.id}`,
    requestActionType: actionTypes.DELETE_ADDRESS,
    requestActionData: {
      addressId: address.id
    }
  }
  return deleteEndpoint(request)
}

export function setAddress (address, formName) {
  return {
    type: actionTypes.SET_ADDRESS,
    address: address,
    formName: formName
  }
}

function newAddressBookEntryPayload (address) {
  return {
    data: {
      type: 'addresses',
      attributes: {
        first_name: address.first_name,
        last_name: address.last_name,
        address_line_1: address.line_1,
        address_line_2: address.line_2,
        city: address.city,
        state: address.state,
        country: address.country_code,
        postcode: address.zipcode,
        preferred_billing: address.preferred_billing,
        preferred_shipping: address.preferred_shipping,
        meta_attributes: addressMetaAttributes(address)
      }
    }
  }
}

function addressMetaAttributes (address) {
  return {
    label: {
      value: address.label
    },
    company_name: {
      value: address.companyName
    },
    phone_number: {
      value: address.primary_phone
    },
    email: {
      value: address.email
    }
  }
}
