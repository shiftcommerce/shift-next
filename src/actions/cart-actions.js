// Libraries
import Cookies from 'js-cookie'

// Lib
import ApiClient from '../lib/api-client'

// Actions
import * as actionTypes from './action-types'
import { readEndpoint, postEndpoint } from './api-actions'

const fetchCartRequest = {
  endpoint: '/getCart',
  successActionType: actionTypes.CART_UPDATED
}

export function readCart (options = {}) {
  return (dispatch, getState) => {
    // Only fetch the cart if needed, otherwise do nothing
    if ((Cookies.get('cart') && !getState().cart.id) || options.force) {
      return dispatch(readEndpoint(fetchCartRequest))
    } else {
      return Promise.resolve()
    }
  }
}

const addToCartRequest = (variantId, quantity) => {
  return {
    endpoint: '/addToCart',
    body: { variantId, quantity },
    successActionType: actionTypes.CART_UPDATED
  }
}

export function addToCart (variantId, quantity) {
  return postEndpoint(addToCartRequest(variantId, quantity))
}

const updateLineItemQuantityRequest = (lineItemId, newQuantity) => {
  return {
    endpoint: '/updateLineItem',
    body: { lineItemId, newQuantity },
    successActionType: actionTypes.CART_UPDATED
  }
}

export function updateLineItemQuantity (lineItemId, newQuantity) {
  return postEndpoint(updateLineItemQuantityRequest(lineItemId, newQuantity))
}

const deleteLineItemRequest = (lineItemId) => {
  return {
    endpoint: `/deleteLineItem/${lineItemId}`,
    successActionType: actionTypes.CART_UPDATED
  }
}

export function deleteLineItem (lineItemId) {
  return postEndpoint(deleteLineItemRequest(lineItemId))
}

const addCartCouponRequest = (couponCode) => {
  return {
    endpoint: '/addCartCoupon',
    body: { couponCode }
  }
}

export const fetchShippingMethodsRequest = () => {
  return {
    endpoint: '/getShippingMethods'
  }
}

const setShippingMethodRequest = (shippingMethodId) => {
  return {
    endpoint: '/setShippingMethod',
    body: { shippingMethodId },
    successActionType: actionTypes.CART_UPDATED
  }
}

export function setCartShippingMethod (shippingMethodId) {
  return postEndpoint(setShippingMethodRequest(shippingMethodId))
}

export async function submitCoupon (couponCode) {
  const request = addCartCouponRequest(couponCode)
  const response = await new ApiClient().post(request.endpoint, request.body)
  if (response.status === 201) {
    return response.data
  }
  throw response.data
}

export function setAPIError (error, setErrors) {
  setErrors({ couponCode: error[0]['title'] })
}

const setBillingAddressRequest = (addressId) => {
  return {
    endpoint: '/setCartBillingAddress',
    body: { addressId },
    successActionType: actionTypes.CART_UPDATED
  }
}

export function setCartBillingAddress (addressId) {
  return postEndpoint(setBillingAddressRequest(addressId))
}

const setShippingAddressRequest = (addressId) => {
  return {
    endpoint: '/setCartShippingAddress',
    body: { addressId },
    successActionType: actionTypes.CART_UPDATED
  }
}

export function setCartShippingAddress (addressId) {
  return postEndpoint(setShippingAddressRequest(addressId))
}

const createShippingAddressRequest = (address) => {
  return {
    endpoint: '/createAddress',
    body: address,
    successActionType: actionTypes.SHIPPING_ADDRESS_CREATED
  }
}

export function createShippingAddress (address) {
  return postEndpoint(createShippingAddressRequest(address))
}

const createBillingAddressRequest = (address) => {
  return {
    endpoint: '/createAddress',
    body: address,
    successActionType: actionTypes.BILLING_ADDRESS_CREATED
  }
}

export function createBillingAddress (address) {
  return postEndpoint(createBillingAddressRequest(address))
}

export function toggleMiniBag (displayed) {
  return {
    type: actionTypes.TOGGLE_MINIBAG,
    displayed
  }
}
