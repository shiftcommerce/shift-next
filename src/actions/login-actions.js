// actionTypes
import * as actionTypes from './action-types'

// Actions
import { postEndpoint } from './api-actions'

export function createLogin (account) {
  const request = {
    endpoint: '/login',
    body: createLoginPayload(account),
    requestActionType: actionTypes.GET_LOGIN,
    errorActionType: actionTypes.ERROR_LOGIN
  }
  return postEndpoint(request)
}

function createLoginPayload (login) {
  return {
    data: {
      type: 'customer_account_authentications',
      attributes: {
        email: login.email,
        password: login.password,
        meta_attributes: {}
      }
    }
  }
}
