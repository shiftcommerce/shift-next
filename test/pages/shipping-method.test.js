// Libraries
import Router from 'next/router'
import nock from 'nock'
import Cookies from 'js-cookie'

// Pages
import ShippingMethodPage from '../../src/pages/checkout/shipping-method'
import * as apiActions from '../../src/actions/api-actions'

// Lib
import Config from '../../src/lib/config'

const shippingAddress = {
  country_code: 'GB',
  first_name: 'First Name',
  last_name: 'Last Name',
  line_1: 'Test House',
  zipcode: 'TEST POSTCODE',
  city: 'Leeds',
  state: 'Yorkshire',
  primary_phone: '01234567890',
  email: 'test@example.com',
  collapsed: false,
  completed: false,
  showEditButton: true,
  errors: {}
}

afterEach(() => { nock.cleanAll() })

jest.mock('next/config', () => () => ({
  publicRuntimeConfig: {}
}))

test('sets paymentMethod in state when instantiated', () => {
  // Arrange
  const cookieSpy = jest.spyOn(Cookies, 'get').mockImplementation(() => 'PayPal')
  const pushSpy = jest.spyOn(Router, 'push').mockImplementation(() => {})
  const cartState = {
    shipping_address: { id: 99 }
  }
  const checkoutState = {}
  const thirdPartyPaymentMethodOptions = ['PayPal']
  const setCurrentStep = jest.fn()

  // Act
  const wrapper = shallow(
    <ShippingMethodPage cart={cartState} checkout={checkoutState} setCurrentStep={setCurrentStep} thirdPartyPaymentMethods={thirdPartyPaymentMethodOptions}/>,
    { disableLifecycleMethods: true }
  )

  // Assert
  expect(wrapper.instance().state.paymentMethod).toBe('PayPal')
  cookieSpy.mockRestore()
  pushSpy.mockRestore()
})

test('componentDidMount() redirects to the shipping address page when one is not set when the default card payment option is used', () => {
  // Arrange
  const pushSpy = jest.spyOn(Router, 'push').mockImplementation(() => {})
  const cartState = {}
  const checkoutState = {}
  const thirdPartyPaymentMethodOptions = ['PayPal']

  // Act
  shallow(<ShippingMethodPage cart={cartState} checkout={checkoutState} thirdPartyPaymentMethods={thirdPartyPaymentMethodOptions}/>)

  // Assert
  expect(pushSpy).toHaveBeenCalledWith('/checkout/shipping-address')
  pushSpy.mockRestore()
})

test('componentDidMount() redirects to the payment method page when shipping address is not set when third party payment is used', () => {
  // Arrange
  const cookieSpy = jest.spyOn(Cookies, 'get').mockImplementation(() => 'PayPal')
  const pushSpy = jest.spyOn(Router, 'push').mockImplementation(() => {})
  const cartState = {
    shipping_address: { id: 99 }
  }
  const checkoutState = {}
  const thirdPartyPaymentMethodOptions = ['PayPal']

  // Act
  shallow(
    <ShippingMethodPage cart={cartState} checkout={checkoutState} thirdPartyPaymentMethods={thirdPartyPaymentMethodOptions}/>
  )

  // Assert
  expect(pushSpy).toHaveBeenCalledWith('/checkout/payment-method')
  cookieSpy.mockRestore()
  pushSpy.mockRestore()
})

test('componentDidMount() redirects to the payment method page when billing address is not set when third party payment is used', () => {
  // Arrange
  const cookieSpy = jest.spyOn(Cookies, 'get').mockImplementation(() => 'PayPal')
  const pushSpy = jest.spyOn(Router, 'push').mockImplementation(() => {})
  const cartState = {}
  const checkoutState = {}
  const thirdPartyPaymentMethodOptions = ['PayPal']

  // Act
  shallow(
    <ShippingMethodPage cart={cartState} checkout={checkoutState} thirdPartyPaymentMethods={thirdPartyPaymentMethodOptions}/>
  )

  // Assert
  expect(pushSpy).toHaveBeenCalledWith('/checkout/payment-method')
  cookieSpy.mockRestore()
  pushSpy.mockRestore()
})

test('renders correct checkout components', () => {
  // Arrange
  const thirdPartyPaymentMethodOptions = ['PayPal']
  const checkoutState = {
    shippingAddress: {
      collapsed: true,
      completed: true,
      showEditButton: true
    },
    paymentMethod: 'PayPal'
  }

  // Act
  const wrapper = shallow(<ShippingMethodPage cart={{ shipping_address: {} }} checkout={checkoutState} thirdPartyPaymentMethods={thirdPartyPaymentMethodOptions} />, {
    disableLifecycleMethods: true
  })

  // Assert
  expect(wrapper).toMatchSnapshot()
  expect(wrapper.find('AddressFormSummary').length).toEqual(1)
  // ShippingMethods component doesn't render as there aren't any
  // shippingMethods returned from the API yet
  expect(wrapper.find('ShippingMethods').length).toEqual(0)
})

test('render shipping methods as expected', async () => {
  // Arrange
  const cookieSpy = jest.spyOn(Cookies, 'get').mockImplementation(() => 'Credit/Debit Card')

  const cart = {
    line_items: [
      {
        id: '1'
      }
    ],
    line_items_count: 1,
    shipping_address: shippingAddress,
    shipping_method: {
      dummy: true
    }
  }

  const checkout = {
    shippingAddress: {
      collapsed: true,
      completed: true,
      showEditButton: true
    },
    shippingMethod: {}
  }

  const fetchShippingSpy = jest.spyOn(ShippingMethodPage, 'fetchShippingMethods').mockImplementation(() => Promise.resolve({
    data: [{
      id: 1,
      sku: 'STA_SHIP',
      label: 'Standard shipping',
      total: 3.75,
      meta_attributes: {
        working_days: {
          value: 2
        }
      }
    }]
  }))

  const thirdPartyPaymentMethodOptions = ['PayPal']

  // Act
  const wrapper = mount(<ShippingMethodPage cart={cart} checkout={checkout} thirdPartyPaymentMethods={thirdPartyPaymentMethodOptions} dispatch={jest.fn()} />)

  // Assert
  expect(wrapper.find('Loading')).toBeTruthy()
  expect(wrapper).toMatchSnapshot()

  await wrapper.instance().componentDidMount()

  expect(wrapper).toIncludeText('1 item')
  expect(wrapper).toIncludeText('Standard shipping')

  fetchShippingSpy.mockRestore()
  cookieSpy.mockRestore()
})

test('renders line item quantity as expected', async () => {
  // Arrange
  const cookieSpy = jest.spyOn(Cookies, 'get').mockImplementation(() => 'Credit/Debit Card')
  const cart = {
    line_items: [],
    line_items_count: 0,
    shipping_address: shippingAddress,
    shipping_method: {
      id: 1,
      sku: 'STA_SHIP',
      label: 'Standard shipping',
      total: 3.75,
      meta_attributes: {
        working_days: {
          value: 2
        }
      }
    }
  }

  const checkout = {
    shippingAddress: {
      collapsed: true,
      completed: true,
      showEditButton: true
    },
    shippingMethod: {
      collapsed: false
    }
  }

  const fetchShippingSpy = jest.spyOn(ShippingMethodPage, 'fetchShippingMethods').mockImplementation(() => Promise.resolve({ data: [] }))

  const thirdPartyPaymentMethodOptions = ['PayPal']

  // Act
  const wrapper = mount(<ShippingMethodPage cart={cart} checkout={checkout} thirdPartyPaymentMethods={thirdPartyPaymentMethodOptions} />)

  // Assert
  expect(wrapper.find('Loading')).toBeTruthy()
  await wrapper.instance().componentDidMount()

  expect(wrapper).toMatchSnapshot()
  expect(wrapper).toIncludeText('0 items')

  fetchShippingSpy.mockRestore()
})

test('preselects first shipping method when fetching shipping methods and none is set', async () => {
  // Arrange
  const cookieSpy = jest.spyOn(Cookies, 'get').mockImplementation(() => 'Credit/Debit Card')
  const cart = {
    line_items: [],
    line_items_count: 0,
    shipping_address: shippingAddress
  }

  const checkout = {
    shippingAddress: {
      collapsed: true,
      completed: true,
      showEditButton: true
    },
    shippingMethod: {
      collapsed: false
    }
  }
  const fetchShippingSpy = jest.spyOn(ShippingMethodPage, 'fetchShippingMethods').mockImplementation(() => Promise.resolve({
    data: [{
      id: 1,
      sku: 'STA_SHIP',
      label: 'Standard shipping',
      total: 3.75,
      meta_attributes: {
        working_days: {
          value: 2
        }
      }
    }, {
      id: 2,
      sku: 'NEX_SHIP',
      label: 'Next day delivery',
      total: 6.25,
      meta_attributes: {
        working_days: {
          value: 1
        }
      }
    }]
  }))

  const postSpy = jest.spyOn(apiActions, 'postEndpoint').mockImplementation(() => { cart.shipping_method = { id: 1 } })

  const thirdPartyPaymentMethodOptions = ['PayPal']

  // Act
  const wrapper = mount(<ShippingMethodPage cart={cart} checkout={checkout} thirdPartyPaymentMethods={thirdPartyPaymentMethodOptions} dispatch={jest.fn()} />)

  // Assert
  expect(wrapper.find('Loading')).toBeTruthy()

  await wrapper.instance().componentDidMount()

  const request = postSpy.mock.calls[0][0]
  expect(request.endpoint).toEqual('/setShippingMethod')
  expect(request.body.shippingMethodId).toEqual(1)

  postSpy.mockRestore()
  fetchShippingSpy.mockRestore()
  cookieSpy.mockRestore()
})

test('selecting a shipping method makes a correct API call', async () => {
  // Arrange
  const cookieSpy = jest.spyOn(Cookies, 'get').mockImplementation(() => 'Credit/Debit Card')
  const cart = {
    line_items: [],
    line_items_count: 1,
    shipping_address: shippingAddress,
    shipping_method: {
      id: 1,
      sku: 'STA_SHIP',
      label: 'Standard shipping',
      total: 3.75,
      meta_attributes: {
        working_days: {
          value: 2
        }
      }
    }
  }

  const checkout = {
    shippingAddress: {
      collapsed: true,
      completed: true,
      showEditButton: true
    },
    shippingMethod: {
      collapsed: false
    }
  }

  const fetchShippingSpy = jest.spyOn(ShippingMethodPage, 'fetchShippingMethods').mockImplementation(() => Promise.resolve({
    data: [{
      id: 1,
      sku: 'STA_SHIP',
      label: 'Standard shipping',
      total: 3.75,
      meta_attributes: {
        working_days: {
          value: 2
        }
      }
    }, {
      id: 2,
      sku: 'NEX_SHIP',
      label: 'Next day delivery',
      total: 6.25,
      meta_attributes: {
        working_days: {
          value: 1
        }
      }
    }]
  }))

  const postSpy = jest.spyOn(apiActions, 'postEndpoint').mockImplementation(() => { cart.shipping_method = { id: 1 } })

  const thirdPartyPaymentMethodOptions = ['PayPal']

  // Act
  const wrapper = mount(<ShippingMethodPage cart={cart} checkout={checkout} thirdPartyPaymentMethods={thirdPartyPaymentMethodOptions} dispatch={jest.fn()} />)

  // Assert
  expect(wrapper.find('Loading')).toBeTruthy()

  await wrapper.instance().componentDidMount()
  wrapper.update()

  wrapper.find('input[id="NEX_SHIP_2"]').simulate('change', { target: { checked: true } })

  const request = postSpy.mock.calls[0][0]
  expect(request.endpoint).toEqual('/setShippingMethod')
  expect(request.body.shippingMethodId).toEqual(2)

  fetchShippingSpy.mockRestore()
  postSpy.mockRestore()
  cookieSpy.mockRestore()
})

test('fetchShippingMethods() returns shipping methods from the API', async () => {
  Config.set({
    apiHostProxy: 'http://example.com'
  })

  nock('http://example.com')
    .defaultReplyHeaders({ 'access-control-allow-origin': '*' })
    .get('/getShippingMethods')
    .reply(200, { data: 'shipping methods data' })

  expect(await ShippingMethodPage.fetchShippingMethods()).toEqual({ data: 'shipping methods data' })
})

test('fetches shipping methods, sorts them by total and puts them in state', async () => {
  // Arrange
  const cookieSpy = jest.spyOn(Cookies, 'get').mockImplementation(() => 'PayPal')
  const shippingMethods = {
    data: [{
      id: 1,
      total: 20,
      meta_attributes: {
        working_days: {
          value: 1
        }
      }
    }, {
      id: 2,
      total: 10,
      meta_attributes: {
        working_days: {
          value: 1
        }
      }
    }]
  }

  const cartState = {
    shipping_address: shippingAddress,
    billing_address: { id: 2 },
    shipping_method: {}
  }

  const checkoutState  = {
    shippingAddress: {
      collapsed: true,
      completed: true,
      showEditButton: true
    },
    shippingMethod: {
      collapsed: false
    }
  }
  const thirdPartyPaymentMethodOptions = ['PayPal']
  const instance = mount(
    <ShippingMethodPage cart={cartState} checkout={checkoutState} thirdPartyPaymentMethods={thirdPartyPaymentMethodOptions}/>
  ).instance()
  const fetchShippingMethodsSpy = jest.spyOn(instance.constructor, 'fetchShippingMethods').mockImplementation(() => shippingMethods)

  // Act
  await instance.componentDidMount()

  // Assert
  expect(instance.state.loading).toEqual(false)
  expect(instance.state.shippingMethods).toEqual(shippingMethods.data)
  expect(instance.state.paymentMethod).toEqual('PayPal')
  fetchShippingMethodsSpy.mockRestore()
  cookieSpy.mockRestore()
})
