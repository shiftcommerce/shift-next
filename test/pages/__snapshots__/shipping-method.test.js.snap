// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render shipping methods as expected 1`] = `
<ShippingMethodPage
  cart={
    Object {
      "line_items": Array [
        Object {
          "id": "1",
        },
      ],
      "line_items_count": 1,
      "shipping_address": Object {
        "city": "Leeds",
        "collapsed": false,
        "completed": false,
        "country_code": "GB",
        "email": "test@example.com",
        "errors": Object {},
        "first_name": "First Name",
        "last_name": "Last Name",
        "line_1": "Test House",
        "primary_phone": "01234567890",
        "showEditButton": true,
        "state": "Yorkshire",
        "zipcode": "TEST POSTCODE",
      },
      "shipping_method": Object {
        "dummy": true,
      },
    }
  }
  checkout={
    Object {
      "shippingAddress": Object {
        "collapsed": true,
        "completed": true,
        "showEditButton": true,
      },
      "shippingMethod": Object {},
    }
  }
  dispatch={[MockFunction]}
  thirdPartyPaymentMethods={
    Array [
      "PayPal",
    ]
  }
>
  <div>
    <div
      className="c-checkout__addressform"
    >
      <div
        className="o-form__address"
      >
        <AddressFormSummary
          city="Leeds"
          firstName="First Name"
          lastName="Last Name"
          onClick={[Function]}
          showEditButton={true}
        >
          <address_form_header_AddressFormHeader
            collapsed={true}
            onClick={[Function]}
            showEditButton={true}
            title="Shipping Address"
          >
            <div
              className="o-form__header c-address-form__header"
            >
              <div
                className="o-form__header-title c-address-form__header-title"
              >
                <h2>
                  Shipping Address
                </h2>
              </div>
              <Button
                className="o-button-edit"
                label="Edit"
                onClick={[Function]}
                status="secondary"
              >
                <button
                  className="o-button-edit o-button o-button--secondary"
                  disabled={false}
                  onClick={[Function]}
                  role="button"
                  tabIndex="0"
                >
                  Edit
                </button>
              </Button>
            </div>
          </address_form_header_AddressFormHeader>
          <div
            className="o-form__wrapper--collapsed c-address-form__summary"
          >
            <p
              className="u-bold"
            >
              First Name
               
              Last Name
               
            </p>
            <span>
              , 
              Leeds
              , 
            </span>
          </div>
        </AddressFormSummary>
      </div>
    </div>
    <Loading>
      <Loading
        className="o-loading"
        color="#f5c6cb"
        delay={250}
        height={50}
        type="spin"
        width={50}
      >
        <div
          className="o-loading"
          dangerouslySetInnerHTML={
            Object {
              "__html": "<svg class=\\"icon-blank\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 32 32\\"></svg>
",
            }
          }
          style={
            Object {
              "fill": "#f5c6cb",
              "height": 50,
              "width": 50,
            }
          }
        />
      </Loading>
    </Loading>
  </div>
</ShippingMethodPage>
`;

exports[`renders correct checkout components 1`] = `
<div>
  <div
    className="c-checkout__addressform"
  >
    <div
      className="o-form__address"
    >
      <AddressFormSummary
        onClick={[Function]}
        showEditButton={true}
      />
    </div>
  </div>
  <Loading />
</div>
`;

exports[`renders line item quantity as expected 1`] = `
<ShippingMethodPage
  cart={
    Object {
      "line_items": Array [],
      "line_items_count": 0,
      "shipping_address": Object {
        "city": "Leeds",
        "collapsed": false,
        "completed": false,
        "country_code": "GB",
        "email": "test@example.com",
        "errors": Object {},
        "first_name": "First Name",
        "last_name": "Last Name",
        "line_1": "Test House",
        "primary_phone": "01234567890",
        "showEditButton": true,
        "state": "Yorkshire",
        "zipcode": "TEST POSTCODE",
      },
      "shipping_method": Object {
        "id": 1,
        "label": "Standard shipping",
        "meta_attributes": Object {
          "working_days": Object {
            "value": 2,
          },
        },
        "sku": "STA_SHIP",
        "total": 3.75,
      },
    }
  }
  checkout={
    Object {
      "shippingAddress": Object {
        "collapsed": true,
        "completed": true,
        "showEditButton": true,
      },
      "shippingMethod": Object {
        "collapsed": false,
      },
    }
  }
  thirdPartyPaymentMethods={
    Array [
      "PayPal",
    ]
  }
>
  <div>
    <div
      className="c-checkout__addressform"
    >
      <div
        className="o-form__address"
      >
        <AddressFormSummary
          city="Leeds"
          firstName="First Name"
          lastName="Last Name"
          onClick={[Function]}
          showEditButton={true}
        >
          <address_form_header_AddressFormHeader
            collapsed={true}
            onClick={[Function]}
            showEditButton={true}
            title="Shipping Address"
          >
            <div
              className="o-form__header c-address-form__header"
            >
              <div
                className="o-form__header-title c-address-form__header-title"
              >
                <h2>
                  Shipping Address
                </h2>
              </div>
              <Button
                className="o-button-edit"
                label="Edit"
                onClick={[Function]}
                status="secondary"
              >
                <button
                  className="o-button-edit o-button o-button--secondary"
                  disabled={false}
                  onClick={[Function]}
                  role="button"
                  tabIndex="0"
                >
                  Edit
                </button>
              </Button>
            </div>
          </address_form_header_AddressFormHeader>
          <div
            className="o-form__wrapper--collapsed c-address-form__summary"
          >
            <p
              className="u-bold"
            >
              First Name
               
              Last Name
               
            </p>
            <span>
              , 
              Leeds
              , 
            </span>
          </div>
        </AddressFormSummary>
      </div>
    </div>
    <Loading>
      <Loading
        className="o-loading"
        color="#f5c6cb"
        delay={250}
        height={50}
        type="spin"
        width={50}
      >
        <div
          className="o-loading"
          dangerouslySetInnerHTML={
            Object {
              "__html": "<svg class=\\"icon-blank\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 32 32\\"></svg>
",
            }
          }
          style={
            Object {
              "fill": "#f5c6cb",
              "height": 50,
              "width": 50,
            }
          }
        />
      </Loading>
    </Loading>
  </div>
</ShippingMethodPage>
`;
